pkg/cel/compatibility.go

```diff
@@ -0,0 +1,236 @@
+	}
+}
+
+// areListTypesCompatible checks if list element types are structurally compatible.
+func areListTypesCompatible(output, expected *cel.Type, provider *apiservercel.DeclTypeProvider) bool {
+	outputParams := output.Parameters()
+	expectedParams := expected.Parameters()
+
+	// Both must have element type parameters
+	if len(outputParams) == 0 || len(expectedParams) == 0 {
+		return len(outputParams) == len(expectedParams)
+	}
+
+	// Recursively check element type compatibility
+	return AreTypesStructurallyCompatible(outputParams[0], expectedParams[0], provider)
+}
+
+// areMapTypesCompatible checks if map key and value types are structurally compatible.
+func areMapTypesCompatible(output, expected *cel.Type, provider *apiservercel.DeclTypeProvider) bool {
+	outputParams := output.Parameters()
+	expectedParams := expected.Parameters()
+
+	// Both must have key and value type parameters
+	if len(outputParams) < 2 || len(expectedParams) < 2 {
+		return len(outputParams) == len(expectedParams)
+	}
+
+	// Check key type compatibility
+	if !AreTypesStructurallyCompatible(outputParams[0], expectedParams[0], provider) {
+		return false
+	}
+
+	// Check value type compatibility
+	return AreTypesStructurallyCompatible(outputParams[1], expectedParams[1], provider)
+}
+
+// areStructTypesCompatible checks if struct types are structurally compatible
+// by introspecting their fields using the DeclTypeProvider.
+func areStructTypesCompatible(output, expected *cel.Type, provider *apiservercel.DeclTypeProvider) bool {
+	if provider == nil {
+		// Without provider, we can't introspect fields - fall back to kind check only
+		return true
+	}
+
+	// Resolve DeclTypes by walking through nested type paths
+	expectedDecl := resolveDeclTypeFromPath(expected.String(), provider)
+	outputDecl := resolveDeclTypeFromPath(output.String(), provider)
+
+	// If we can't resolve both types, we can't do structural comparison
+	// Fall back to accepting it (permissive - could make this stricter)
+	if expectedDecl == nil || outputDecl == nil {
+		return true
+	}
+
+	// Check that output has all required fields of expected
+	return areStructFieldsCompatible(outputDecl, expectedDecl, provider)
+}
+
+// resolveDeclTypeFromPath resolves a DeclType by walking through a nested path.
+// For example, "__type_ingressroute.spec.routes.@idx.middlewares" would:
+// 1. Strip __type_ prefix and look up "ingressroute" in the provider
+// 2. Find the "spec" field
+// 3. Find the "routes" field
+// 4. Get the list element type (@idx)
+// 5. Find the "middlewares" field
+func resolveDeclTypeFromPath(typePath string, provider *apiservercel.DeclTypeProvider) *apiservercel.DeclType {
+	if provider == nil || typePath == "" {
+		return nil
+	}
+
+	// Split the path into segments
+	segments := strings.Split(typePath, ".")
+	if len(segments) == 0 {
+		return nil
+	}
+
+	// Get the root name, stripping __type_ prefix if present
+	rootName := segments[0]
+	if strings.HasPrefix(rootName, "__type_") {
```

Comment:
nit: this seems like a hack thats based on your previous type prefixing, maybe this should become a const or things will break